pkg/network/openshift_sdn.go

```diff
@@ -52,6 +53,9 @@ func renderOpenShiftSDN(conf *operv1.NetworkSpec, manifestDir string) ([]*uns.Un
 	}
 	data.Data["NodeConfig"] = nodeCfg
 
+	data.Data["CNIConfig"] = makeCNIConfig(conf, CNINetworkName, "0.3.1",
+		`{"type": "openshift-sdn"}`)
```

Comment:
Not super happy about this, can we just have the SDN itself do the combination? eg the CNO makes the template from the ChainedPlugins and the SDN inserts itself at position 0 in the list and then writes the result conflist to the final CNI configdir that CRIO looks for?

Reply:
I was going to say something similar before, but there is already a crazy amount of coupling between the operator and the SDN plugin itself anyway. This doesn't really make it worse. (And besides, what happens if one of the chained plugins wants to use some feature that requires using version 0.4.0 of the spec?)

Reply:
Yeah, I thought about that for a long time, and couldn't find a solution I like.

First of all, the cniVersion is a tricky coordination point no matter where you place it. It needs to be the same for all plugins, and all plugins need to support the selected versions. If we bump openshift-sdn to 0.4.0, then all chained plugins instantly *must* support 0.4.0. I suppose this is a problem with CNI, and we may wish to revisit that decision. Wherever we do the config munging doesn't fix this problem.

As for having openshift-sdn insert itself, I decided against that for a few reasons:

1. In general, manipulating untyped json is a pain in go :-)
2. This is easier add to other plugins, e.g. ovn. We can just write a sidecar in bash that waits for healthz and copies a file.
3. Generally, CNI configuration is expected to be administrator-defined. openshift-sdn writes it's own configuration file for availability signalling and historical reasons. And because it has no knobs.

I'm happy to revisit this. It's not great either way you look at it.